### PR TITLE
prep_char_num_sort, prep_num_band, expl_cond_dist_tbl

### PR DESCRIPTION
**Pull request by CRSlade1**
Finally sorted these functions out. Also didn't rebase as I think I now understand what to do to fix a conflict. 

Also would be curious to hear thoughts on leaving #TODO: statements at the end of functions that we don't think warrants raising an issue.

### CODE REVIEW

File: R/prep_char_num_sort.R [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  #'
+  #' @section Inputs:
+  #' @param array_to_sort Array[Character] - This is the array which we want to sort. (If you create c(1,2,"3") they all get assigned character.)
+  #' @param NA_val Character/Numeric/NA (Default = "NA") - This is the value which we want to put NAs as.
+  #'   I recommend using "_NA" when you need to treat NAs as a character to get it positioned near the start rather than in the middle which "NA" would do.
+  #'   If NA is not NA and cannot be converted to numeric, NA will get treated as character and therefore get affected by char_decreasing/
+  #' @param char_decreasing Boolean/NULL (Default = FALSE) - Set this to TRUE if you want character values sorted into descending order. Set to NULL if you do not want character values sorted.
+  #' @param num_decreasing Boolean/NULL (Default = FALSE) - Set this to TRUE if you want numeric values sorted into descending order. Set to NULL if you do not want numeric values sorted.
+  #' @param unique Boolean (Default = FALSE) - Set this to true if you want unique sorted outputs after handling NAs.
```
Comment (clementstross) [inline R/prep_char_num_sort.R:15]: I think the default should be true

File: R/prep_char_num_sort.R [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  #' @examples
+  #' prep_char_num_sort(c("100","-100","1","-1","_","-","''","CAT","LLAMA","NA","NULL",NA,NULL),NA_val = 2)
+  #' # Output: c("''","-","_","CAT","LLAMA","NA","NULL","-100","-1","1","2","100")
+  #' prep_char_num_sort(c("C","M","T","-55555","-44444","-33333","0",NA),NA_val = "-1")
+  #'
+  #' @export
+
+
+  # Checking that char_decreasing, num_decreasing and unique are all logical (or NULL for the decreasing options)
+  checkmate::assert_logical(char_decreasing,null.ok = TRUE)
```
Comment (clementstross) [inline R/prep_char_num_sort.R:32]: len=1?

File: R/prep_num_band.R [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  #'
+  #' prep_num_band(var_to_band=c(-100, 200, 0, 1.5, 3.14159, 10, 7, -200), bands=c(-150, -10, 4, 5), out_of_bands = 999)
+  #' # Output: (-10, 999, 4, 4, 4, 999, 999, -150)
+  #'
+  #' @export
+
+
+  # Checking that bands and out_of_bands are numeric. Any.missing checks for NAs.
+  checkmate::assert_numeric(bands, any.missing = FALSE)
+  checkmate::assert_numeric(out_of_bands,any.missing = FALSE)
```
Comment (clementstross) [inline R/prep_num_band.R:34]: len=1

File: R/prep_char_num_sort.R [hunk 4]
```diff
@@ -0,0 +1,96 @@
+prep_char_num_sort <- function(array_to_sort, NA_val = NA, char_decreasing=FALSE, num_decreasing=FALSE, unique=FALSE){
```
Comment (clementstross) [inline R/prep_char_num_sort.R:1]: This will still get file_10 and file_2 the wrong way round.

File: tests/testthat/test_prep_char_num_sort.R [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  testthat::expect_equal(prep_char_num_sort(c("G",100,"A",1,2,50,NA),NA_val = NA),c("A","G","1","2","50","100",NA))
+
+
+  # Check that it works with only numeric values and returns a numeric output (if NA_val is not a string).
+  testthat::expect_equal(prep_char_num_sort(c(1,100,2,50,NA),NA_val = NA),c(1,2,50,100,NA))
+  testthat::expect_equal(prep_char_num_sort(c(1,100,2,50,NA),NA_val = 2),c(1,2,2,50,100))
+  testthat::expect_equal(prep_char_num_sort(c(1,100,2,50,NA),NA_val = "A"),c("A","1","2","50","100"))
+
+  # NA_val being character with no NAs, should be numeric.
+  testthat::expect_equal(prep_char_num_sort(c(1,100,2,50),NA_val = "A"),c(1,2,50,100))
```
Comment (clementstross) [inline tests/testthat/test_prep_char_num_sort.R:24]: I don't like this behaviour. The output class changes depending on if there is an NA in the dataset. That could get very confusing.

File: tests/testthat/test_prep_num_band.R [hunk 6]
```diff
@@ -0,0 +1,63 @@
+testthat::context("Testing prep_num_band - A numeric banding function")
+
+testthat::test_that("Testing properties",{
+
+  # Can handle characters
+  testthat::expect_equal(prep_num_band(c("A","B","1","2","C","5","6","7"),c(2,8)),c("A","B","2","2","C","8","8","8"))
```
Comment (clementstross) [inline tests/testthat/test_prep_num_band.R:6]: I think we want a warning or an error here. I don't think letting people band with characters is what we want

File: R/expl_cond_dist_tbl.R [hunk 7]
```diff
@@ -0,0 +1,184 @@
+expl_cond_dist_tbl <- function(main_var, cond_var, output_var="thin", NA_val="_NA", warn_high_band = 50L, err_high_band = 100L, verbose=TRUE){
```
[the remark this answers is not in the document]
Comment (CRSlade1) [reply R/expl_cond_dist_tbl.R:1]: Yep, above makes sense - didn't even realise I did that.

File: R/expl_cond_dist_tbl.R [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+      if (sum(cond_var == NA_val, na.rm = TRUE) > 0) {
+        print(paste0("There are ", sum(cond_var == NA_val, na.rm = TRUE),
+                     " pre-existing values in cond_var equal to ", NA_val, "."))
+      }
+    }
+    cond_var[is.na(cond_var)] <- NA_val
+  }
+
+  # Creating a neat order to display the character and numeric values.
+  sort_main <- prep_char_num_sort(unique(as.character(main_var)))
```
[the remark this answers is not in the document]
Comment (CRSlade1) [reply R/expl_cond_dist_tbl.R:99]: Think it's worth adding an option for including the order?

File: R/expl_cond_dist_tbl.R [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+  }
+  if (length(sort_cond) >= err_high_band) {
+    warning("ERROR: Too many bands in cond_var")
+  }
+  checkmate::assert(length(sort_main) < err_high_band)
+  checkmate::assert(length(sort_cond) < err_high_band)
+
+  # Creating a table of proportions.
+  df <- data.frame(main_var, cond_var)
+  output <- df %>%
```
Comment (CRSlade1) [inline R/expl_cond_dist_tbl.R:115]: I also don't like this variable name - might change this to thin_tbl

File: R/expl_cond_dist_tbl.R [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    dplyr::mutate(prop = n*100 / total) %>%
+    dplyr::arrange(base::match(main_var, sort_main), base::match(cond_var, sort_cond))
+
+  # Group_by & Count: Creates the variable n, just a count of how many in each combination of (main_var, cond_var)
+  # 2nd Group_by & Mutate: Total is just count of those in each level of cond_var.
+  # 3rd Group_by & Mutate: Regroup to level of (main_var, cond_var) and create proportion of cond_var.
+  # Arrange & Match: Reorder table sort_main and sort_cond.
+
+  # Creating wide table that we use for reports.
+  wide_tbl <- data.frame(output) %>%
```
[the remark this answers is not in the document]
Comment (CRSlade1) [reply R/expl_cond_dist_tbl.R:130]: Good question. I don't know.

I think my reasoning at the time was that count is similar and it makes sense to make them look the same so you can compare but might as well just output wide_tbl

File: R/prep_char_num_sort.R (hunk 4 above)
Comment (CRSlade1) [reply R/prep_char_num_sort.R:1]: Yeah you're right. Hadn't even considered this but that's probably a case worth dealing with.  
Unsure how to go about this though because it also feels like I should generalise for the case of being able to deal with folder_1_subfolder_2_file_3 etc.. So split string into array of [chr, num, chr, num, chr, num], sort each of the groups of components and reassemble. Then there's also a whole element of what if not everything has the same structure. Could imagine a case of file2019_v1, file2019_v1.1 etc.

File: R/prep_char_num_sort.R (hunk 1 above)
Comment (CRSlade1) [reply R/prep_char_num_sort.R:15]: I initially had this as TRUE, as I could imagine this mainly being used for visualising results etc which is when I would expect it to be true. Arguably shouldn't even be including a unique function but only did so because of the case of potentially setting the NAs to the same as a pre-existing value.

File: R/expl_cond_dist_tbl.R [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+  }
+  checkmate::assert(length(main_var) == length(cond_var))
+
+  # Check that variables are not NULL.
+  checkmate::assert(!is.null(main_var))
+  checkmate::assert(!is.null(cond_var))
+  checkmate::assert(!is.null(NA_val))
+
+  checkmate::assert_logical(verbose, null.ok = FALSE)
+  checkmate::assert_integerish(warn_high_band, any.missing = FALSE, lower = 2)
```
[the remark this answers is not in the document]
Comment (CRSlade1) [reply R/expl_cond_dist_tbl.R:57]: Yep

File: tests/testthat/test_prep_num_band.R (hunk 6 above)
Comment (CRSlade1) [reply tests/testthat/test_prep_num_band.R:6]: Generally, I agree that this isn't great practice. I think there could however be benefit in Credit Data where you could need to check through loads of variables and most of them have C,M or T levels as well as numerics with potentially large ranges like (-300,500). I could imagine a case where it would be useful to band the levels together and plot them whilst also keeping the C,M and T fields to one side.

File: tests/testthat/test_prep_char_num_sort.R (hunk 5 above)
Comment (CRSlade1) [reply tests/testthat/test_prep_char_num_sort.R:24]: I'm not sure what the best behaviour here would be. I agree that it is a weird thing to impute a series of numeric values to a character all of a sudden but you have had to manually specify that you want to impute to a character. Perhaps throwing a warning?

File: R/expl_cond_dist_tbl.R (hunk 7 above)
[the remark this answers is not in the document]
Comment (clementstross) [reply R/expl_cond_dist_tbl.R:38]: I don't think it matters too much but we just need to decide if we care

File: R/expl_cond_dist_tbl.R (hunk 10 above)
[the remark this answers is not in the document]
Comment (clementstross) [reply R/expl_cond_dist_tbl.R]: Again I don't think it really matters but it makes the code quite hard to follow.

File: R/expl_cond_dist_tbl.R (hunk 8 above)
[the remark this answers is not in the document]
Comment (clementstross) [reply R/expl_cond_dist_tbl.R:98]: Same. You just want the prep_char_num_sort to sort by the factor order. I don't know if anyone other than me uses ordered factors much but it means you can keep useful labels and have %>% as.numeric() %>% as.matrix turn a human-readable data.frame into an xgboost DMatrix

File: R/prep_char_num_sort.R (hunk 4 above)
Comment (clementstross) [reply R/prep_char_num_sort.R]: I agree. More effort than it is worth but I'd maybe leave a comment in the documentation. I'd also look at the case where it matches the regex string \w*\d+\w* and have it sort this case as it is easy. 

- check the string text is the same
- extract \d+
- sort these and use the order to sort the words

If you were really smart you could do the above as a function and call it from inside itself when needed and it could work on all cases

File: tests/testthat/test_prep_char_num_sort.R (hunk 5 above)
Comment (clementstross) [reply tests/testthat/test_prep_char_num_sort.R:24]: Always return a string? For the case of sorting numbers, why would we use your function? For the generic case for giving the order of factor levels, we probably want it to output a sting anyway as it is clearer

File: tests/testthat/test_prep_num_band.R (hunk 6 above)
Comment (clementstross) [reply tests/testthat/test_prep_num_band.R:6]: OK

File: R/expl_cond_dist_tbl.R (hunk 9 above)
Comment (clementstross) [reply R/expl_cond_dist_tbl.R]: I agree

File: R/prep_char_num_sort.R (hunk 4 above)
Comment (CRSlade1) [reply R/prep_char_num_sort.R]: Yeah, this is roughly what I was thinking. Did consider recursion, it's a lovely case for it - I'll have a short think about it but don't expect anything!

File: tests/testthat/test_prep_char_num_sort.R (hunk 5 above)
Comment (CRSlade1) [reply tests/testthat/test_prep_char_num_sort.R:24]: Yep you're right - forgot we can ignore all numeric case. I'll just enforce strings out

File: tests/testthat/test_prep_char_num_sort.R [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+  testthat::expect_equal(prep_char_num_sort(c("A","B","1",NA,NA,"3","C","2")),c("A","B","C","1","2","3",NA,NA))
+
+  # Check that it does sort multidigit numbers. If this did not work, we would see 1,100,2,50 by taking alphabetical with digit size.
+  testthat::expect_equal(prep_char_num_sort(c(1,100,2,50,NA),NA_val = NA),c(1,2,50,100,NA))
+  testthat::expect_equal(prep_char_num_sort(c("G",100,"A",1,2,50,NA),NA_val = NA),c("A","G","1","2","50","100",NA))
+
+
+  # Check that it works with only numeric values and returns a numeric output (if NA_val is not a string).
+  testthat::expect_equal(prep_char_num_sort(c(1,100,2,50,NA),NA_val = NA),c(1,2,50,100,NA))
+  testthat::expect_equal(prep_char_num_sort(c(1,100,2,50,NA),NA_val = 2),c(1,2,2,50,100))
```
Comment (gloverd2) [inline tests/testthat/test_prep_char_num_sort.R:20]: Error: prep_char_num_sort(c(1, 100, 2, 50, NA), NA_val = 2) not equal to c(1, 2, 2, 50, 100).
Lengths differ: 4 is not 5

unqiue issue again

File: tests/testthat/test_prep_char_num_sort.R [hunk 13]
```diff
@@ -0,0 +1,110 @@
+testthat::context("Testing prep_char_num_sort")
+
+
+testthat::test_that("Testing properties",{
```
Comment (gloverd2) [inline tests/testthat/test_prep_char_num_sort.R:4]: This testing segment is very long . where you have used comments you could be breaking them up . if it errors it returns the label of the test ("_Test failed_ 'Testing properties'"). so to break them up makes debugging much easier.

File: tests/testthat/test_expl_cond_dist_tbl.R [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+  # Working example
+  prop_df <- data.frame(main_var = c("1","2","3"), X1 = c(80,0,20), X2 = c(25,75,0))
+  colnames(prop_df) <- c("main_var",1,2)
+  testthat::expect_equal(expl_cond_dist_tbl(main_var = c(1,1,2,1,1,1,3,2,2), cond_var = c(1,1,2,2,1,1,1,2,2),output_var = "prop"), prop_df)
+
+  # Testing output_var
+  thin_df <- data.frame(main_var = c(1,1,2), cond_var = c(1,2,1), n = c(3L,1L,1L), total = c(4L,1L,4L), prop = c(75,100,25))
+  row.names(thin_df) <- c(1L,2L,3L)
+  testthat::expect_equal(expl_cond_dist_tbl(main_var = c(1,1,1,2,1), cond_var = c(1,1,1,1,2), output_var = "thin"), thin_df)
+
```
Comment (gloverd2) [inline tests/testthat/test_expl_cond_dist_tbl.R:14]: not equal . returned a df and a tibble

Comment (gloverd2) [reply tests/testthat/test_expl_cond_dist_tbl.R:14]: reoccuring theme across all tests?

**Review by gloverd2 — COMMENTED** [review]
i pulled the latest version on sladec2 and haven't been able to run the tests due to unique = TRUE & the test_expl_cond_dist_tbl giving tibble vs df

**Review by gloverd2 — COMMENTED** [review]
don't add any other functions to this pull request.
prep num band and prep_char_num_sort are much closer to being approved

File: tests/testthat/test_expl_cond_dist_tbl.R (hunk 14 above)
Comment (CRSlade1) [reply tests/testthat/test_expl_cond_dist_tbl.R:14]: I need to adjust these. I was having issues getting this consistent. I think some of it is related to tidyverse version discrepancies between us

File: R/prep_num_band.R [hunk 15]
```diff
@@ -0,0 +1,74 @@
+#' prep_num_band
+#' @description
+#' This can band numeric values with a purely numeric vector or a vector of mixed entries.
+#' Bands are inclusive in the upper bound.
+#' An extra band of 99,999,999 will be added if there are values above the largest band.
+#' e.g. If I have a value of 500 and bands of 200, 500 and 800: 500 will be in the 500 band but 500.01 would be in 800.
+#'
+#' Note: Banding is consistent with how Pricing view bands in rate tables.
```
Comment (clementstross) [inline R/prep_num_band.R:8]: I think I don't understand how pricing band rate tables. I hate this! I'm used to the other way around. Just to check if we have groups
0
1
2
3
I'd have thought 1.5 gets the rate of 1 but you're telling me it gets the rate of 2?

File: R/prep_num_band.R [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+  # Checking that the amount of values to be banded is more than 0.
+  checkmate::assert_true(!is.null(var_to_band))
+
+  # Warning for if only 1 unique value.
+  if (length(unique(var_to_band)) == 1) {
+    warning("WARNING: Only 1 unique value")
+  }
+
+  # This is getting the numeric parts of the array.
+  var_to_band_num <- var_to_band[!is.na(suppressWarnings(as.numeric(var_to_band)))]
```
Comment (clementstross) [inline R/prep_num_band.R:49]: Just assert that this is numeric. We don't want stray characters

File: R/prep_num_band.R [hunk 17]
```diff
@@ -0,0 +1,74 @@
+#' prep_num_band
+#' @description
+#' This can band numeric values with a purely numeric vector or a vector of mixed entries.
+#' Bands are inclusive in the upper bound.
```
Comment (clementstross) [inline R/prep_num_band.R:4]: I'd like to make it such that this function and other banding function produced the same output. A list with "bins" and "labels" also have a choice to output the bin index or the bin label

**Review by clementstross — CHANGES_REQUESTED** [review]
Fails the following test
prep_char_num_sort(c("0011","2","3","4","5", NA))

the output given is 
"_NA"  "2"    "3"    "4"    "5"    "1111"

if I remove the NA the problem goes away. I think the issue is in the regex statement

Also needs devtools::document() running check_all_identical.Rd doesn't exist

File: R/prep_char_num_sort.R [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+
+  # Basic assertions
+  checkmate::assert_logical(char_decreasing, null.ok = TRUE)
+  checkmate::assert_logical(num_decreasing, null.ok = TRUE)
+  checkmate::assert_logical(unique_val, null.ok = FALSE)
+  checkmate::assert(!identical(NA_val,TRUE))  # Need identicality here because TRUE == 1
+  checkmate::assert(!identical(NA_val,FALSE))
+  checkmate::assert(length(NA_val) == 1)
+  checkmate::assert(length(array) > 0)
+
```
[the remark this answers is not in the document]
Comment (CRSlade1) [reply R/prep_char_num_sort.R:61]: Oops, forgot to take this out! Was using this for debugging

File: R/check_all_identical.R [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+  if (class(input) != "list") {
+    warning("ERROR: Input MUST be a list. This is not an array. To convert an array to list use as.list()")
+  }
+  # Need to enforce a list. Lists can handle multiple types whereas arrays will force the same. Using c(2,"2") will convert both to numeric in an array but NOT in a list.
+  checkmate::assert(class(input) == "list")
+  # When only one thing remaining, no comparison to be made.
+  if (length(input) == 1) {
+    return(TRUE)
+  } else {
+    # Check that list object 1 and list object 2 are identical. Use an and condition with that an identical check of all objects except 1.
```
[the remark this answers is not in the document]
Comment (CRSlade1) [reply R/check_all_identical.R:31]: It is in the description but sure will leave it here too

File: R/expl_cond_dist_tbl.R [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+  checkmate::assert(length(sort_cond) < err_high_band)
+
+  # Creating a table of proportions.
+  df <- data.frame(main_var, cond_var)
+  thin_tbl <- df %>%
+    dplyr::group_by(main_var, cond_var) %>%
+    dplyr::count() %>%
+    dplyr::group_by(cond_var) %>%
+    dplyr::mutate(total = sum(n)) %>%
+    dplyr::group_by(main_var, cond_var) %>%
```
[the remark this answers is not in the document]
Comment (CRSlade1) [reply R/expl_cond_dist_tbl.R:119]: Didn't know that was a thing

File: R/prep_num_band.R (hunk 15 above)
Comment (CRSlade1) [reply R/prep_num_band.R:8]: Yep. Don't hate the messenger. You'll see in a lot of rate tables that there is 999999 because "they" round up

File: R/prep_char_num_sort.R [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+  } else {
+    # Find first character or numeric chunk. i.e. A_1_A would get A_.
+    first_element <- stringr::str_extract(array,"(\\-?[:digit:]+(\\.[:digit:]+)?)|([A-z\\.\\'@£$%~#&\\!\\?\\- ]+)")
+    # Conditions on NA_val:
+    # If character, needs to be added and sorted with characters.
+    # If numeric, needs to be added and sorted with numbers.
+    # If NA, needs to be added separately as NA.
+
+    # The case of having NAs with NA_val as NA is started here and carried separately.
+    first_nas <- c(if ((is.na(NA_val)) & sum(is.na(array)) > 0) {rep(NA_val,sum(is.na(array)))} else {NULL})
```
[the remark this answers is not in the document]
Comment (CRSlade1) [reply R/prep_char_num_sort.R:77]: Unfortunately this doesn't work as ifelse executes row wise so does not accept array outputs.
E.g. ifelse(TRUE==TRUE, c(1,1,1), NA) gives just 1.

File: R/expl_cond_dist_tbl.R [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+  # Arrange: This is to change rows into sort_main. (Sorted version of main_var)
+  # Mutate_at: Tidying all NAs with 0, in columns that are in sort_cond. This prevents row names also being imputed from NA.
+
+  if (verbose) {
+    print(wide_tbl)
+  }
+
+  # Warning for if bandings are large.
+  # This warning is placed here so that it appears underneath the large table.
+  if (length(sort_main) >= warn_high_band & verbose) {
```
[the remark this answers is not in the document]
Comment (CRSlade1) [reply R/expl_cond_dist_tbl.R:148]: There aren't the same. One of these is for main_var, one is for cond_var.

[the remark this answers is not in the document]
Comment (CRSlade1) [reply R/expl_cond_dist_tbl.R:148]: Although having said that, I also enforced that the were the same length. I've since removed this as it now works without that constraint

File: R/prep_num_band.R (hunk 15 above)
Comment (clementstross) [reply R/prep_num_band.R:8]: I hate that!

File: R/expl_cond_dist_tbl.R (hunk 22 above)
[the remark this answers is not in the document]
Comment (clementstross) [reply R/expl_cond_dist_tbl.R:148]: How does it work without constraints? I'd have thought they'd need to be the same length

File: tests/testthat/test_prep_char_num_sort.R [hunk 23]
```diff
@@ -0,0 +1,121 @@
...
+  testthat::expect_equal(prep_char_num_sort(c("B","B","A","A","3","3","1","2"),unique_val = TRUE),c("A","B","1","2","3"))
+  # unique_val FALSE
+  testthat::expect_equal(prep_char_num_sort(c("B","B","A","A","3","3","1","2"),unique_val = FALSE),c("A","A","B","B","1","2","3","3"))
+
+
+  # Test on NAs.
+  testthat::expect_identical(prep_char_num_sort(c(NA,NA,NA),unique_val = TRUE),as.logical(NA))
+
+  # Negative numbers.
+  testthat::expect_equal(prep_char_num_sort(c("C","M","T","-55555","-44444","-33333","0",NA),NA_val = "-1"),c("-1","C","M","T","-55555","-44444","-33333","0"))
```
Comment (clementstross) [inline tests/testthat/test_prep_char_num_sort.R:65]: These two tests can't be right. The order changes depending on NA_value="-1" vs NA_value=-1. Also I don't like All NAs being put at the start.

What happens with rep_char_num_sort(c("1","2","3",NA),NA_val = "2")
I think we want c("1", "2", "2", "3") i think yours gives c("2", "1",, "2", "3")

File: R/prep_num_band.R (hunk 16 above)
Comment (CRSlade1) [reply R/prep_num_band.R:49]: Discussed this in codebase conversation. Agreed that generally not great but for the sake of Credit Data, it makes everyone's lives easier.

**Review by clementstross — APPROVED** [review]
